ffi/examples/read-table/schema.h

```diff
@@ -0,0 +1,259 @@
+
+/**
+ * This module defines a very simple model of a schema, used only to be able to print the schema of
+ * a table. It consists of a "SchemaBuilder" which is our user data that gets passed into each visit_x
+ * call. This simply keeps track of all the lists we are asked to allocate.
+ *
+ * Each list is a "SchemaItemList", which tracks its length an an array of "SchemaItem"s.
+ *
+ * Each "SchemaItem" has a name and a type, which are just strings. It can also have a list which is
+ * its children. This is initially always NULL, but when visiting a struct, map, or array, we point
+ * this at the list specified in the callback, which allows us to traverse the schema when printing
+ * it. Note that this points to one of the lists in the builder's set of lists and is not a copy.
+ */
+
+// If you want the visitor to print out what it's being asked to do at each step, uncomment the
+// following line
+// #define PRINT_VISITS
+
+typedef struct SchemaItemList SchemaItemList;
+
+typedef struct {
+  char* name;
+  char* type;
+  SchemaItemList* children;
+} SchemaItem;
+
+typedef struct SchemaItemList {
+  uint32_t len;
+  SchemaItem* list;
+} SchemaItemList;
+
+typedef struct {
+  int list_count;
+  SchemaItemList* lists;
+} SchemaBuilder;
+
+char* allocate_name(const KernelStringSlice slice) {
+  char* buf = malloc(sizeof(char) * (slice.len + 1)); // +1 for null
+  snprintf(buf, slice.len + 1, "%s", slice.ptr);
+  return buf;
+}
+
+// lists are preallocated to have exactly enough space, so we just fill in the next open slot and
+// increment our length
+SchemaItem* add_to_list(SchemaItemList *list, char* name, char* type) {
+  int idx = list->len;
+  list->list[idx].name = name;
+  list->list[idx].type = type;
+  list->len++;
+  return list->list+idx;
+}
+
+// print out all items in a list, recursing into any children they may have
+void print_list(SchemaItemList *list, int indent, bool parent_on_last) {
+  for (int i = 0; i < list->len; i++) {
+    bool is_last = i == list->len - 1;
+    for (int j = 0; j <= indent; j++) {
+      if (j == indent) {
+        if (is_last) {
+          printf("└");
+        } else {
+          printf("├");
+        }
+      } else {
+        if (parent_on_last && j == indent - 1) {
+          // don't print a dangling | on my parent's last item
+          printf("   ");
+        } else {
+          printf("│  ");
+        }
+      }
+    }
+    printf("─ %s: %s\n", list->list[i].name, list->list[i].type);
+    if (list->list[i].children) {
+      print_list(list->list[i].children, indent+1, is_last);
+    }
+  }
+}
+
+// declare all our visitor methods
+uintptr_t make_field_list(void *data, uintptr_t reserve) {
+  SchemaBuilder *builder = (SchemaBuilder*)data;
+  int id = builder->list_count;
+#ifdef PRINT_VISITS
+  printf("Making a list of lenth %i with id %i\n", reserve, id);
+#endif
+  builder->list_count++;
+  builder->lists = realloc(builder->lists, sizeof(SchemaItemList) * builder->list_count);
```

Comment:
Yeah, real bad on the pointers 🤦🏽 . 

I've updated things to use indexes.

> Why realloc, out of curiosity? What previous usage could there have been?

We don't know the total number of lists ahead of time. `SchemaBuilder.lists` is an array of `SchemaItemList`, so this realloc is just adding one more to it. Previous usage was when we made the builder and did
```c
  SchemaBuilder builder = {
    .list_count = 0,
    .lists = calloc(0, sizeof(SchemaItem*)),
  };
```

as well as each other call to this method.

> (if there was some valid previous usage, wouldn't we need to free the lists it contained?)

`realloc` does that for you (afaiu): "Unless ptr is NULL, it must have been returned by an earlier call to malloc or related functions.  If the area pointed to was moved, a free(ptr) is done."

Reply:
> > (if there was some valid previous usage, wouldn't we need to free the lists it contained?)
>
> realloc does that for you (afaiu)

I was referring to the lists of `SchemaItem` that each `SchemaItemList` in the builder referred to. But it's not a problem as you say, because we're doing realloc to extend rather than replace.

Reply:
One gotcha with using realloc this way -- it has `O(n**2)` cost, where `n` is the final `list_count`. The usual workaround is to increase the capacity of the array by a fixed ratio like 3/2; in return for potentially 50% wasted space, it is amortized `O(n)` cost. Probably doesn't matter for this example code tho.

Reply:
Yep. I considered increasing more each time, but:
a) it requires tracking an extra capacity param and
b) for example code as you say it's probably overkill